### PR TITLE
implementation of asynchronous metrics collector for better api performance

### PR DESCRIPTION
**Pull request by gs-gunjan**
implementation of asynchronous metrics collector for better api performance

### CODE REVIEW

File: legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/api/QueryMetricsRegistry.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.util.Date;
+
+public interface QueryMetricsRegistry
+{
+    void record(String groupId, String artifactId, String versionId);
+
+    void record(String groupId, String artifactId, String versionId, Date date);
+
+    VersionQueryMetric getFirstInQueue();
```
Comment (elopezcastro) [inline legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/api/QueryMetricsRegistry.java:28]: why do we need get first and find first  in queue... u are exposing your implementation
the methods can be consolidated into one methof findFirst which returns an optional and u deal with that from where u called it

File: legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/services/InMemoryQueryMetricsRegistry.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        this.metrics = new ConcurrentLinkedQueue<>();
+    }
+
+    public void record(String groupId, String artifactId, String versionId)
+    {
+        metrics.offer(new VersionQueryMetric(groupId, artifactId, versionId));
+    }
+
+    public void record(String groupId, String artifactId, String versionId, Date date)
```
Comment (elopezcastro) [inline legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/services/InMemoryQueryMetricsRegistry.java:42]: why do we need a record with a date.? if we do you should implement the above based on this one

File: legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/services/InMemoryQueryMetricsRegistry.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+import org.finos.legend.depot.store.admin.domain.metrics.VersionQueryMetric;
+import org.finos.legend.depot.store.metrics.api.QueryMetricsRegistry;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Date;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+@Singleton
```
Comment (elopezcastro) [inline legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/services/InMemoryQueryMetricsRegistry.java:26]: are u sure this is singleton. every tuiome u create a new project servce impl or any service that requires this which is evry time an api call is amde this will be created as per the binder provided in the module.
you have to bind a singleton differently

File: legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/services/InMemoryQueryMetricsRegistry.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        metrics.offer(new VersionQueryMetric(groupId, artifactId, versionId));
+    }
+
+    public void record(String groupId, String artifactId, String versionId, Date date)
+    {
+        metrics.offer(new VersionQueryMetric(groupId, artifactId, versionId, date));
+    }
+
+    @Override
+    public VersionQueryMetric getFirstInQueue()
```
Comment (elopezcastro) [inline legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/services/InMemoryQueryMetricsRegistry.java:48]: why is this needed... u are expsing the implementation..

File: legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/services/QueryMetricsHandler.java [hunk 5]
```diff
@@ -70,6 +73,14 @@ public List<VersionQueryMetric> findSnapshotVersionMetricsBefore(Date date)
         return metricsStore.findMetricsBefore(date).parallelStream().filter(metric -> VersionValidator.isSnapshotVersion(metric.getVersionId())).collect(Collectors.toList());
     }
 
+    public void persistMetrics()
+    {
+        while (queryMetricsRegistry.findFirst() != null)
```
Comment (elopezcastro) [inline legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/services/QueryMetricsHandler.java:78]: how is this going to work in a concurrent enviorment where the queue is being modified while this is happening

File: legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/api/QueryMetricsRegistry.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+    VersionQueryMetric findFirst();
```
Comment (elopezcastro) [inline legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/api/QueryMetricsRegistry.java:30]: this should return an optional

File: legend-depot-store-api/src/main/java/org/finos/legend/depot/store/admin/api/metrics/QueryMetricsStore.java [hunk 7]
```diff
@@ -27,7 +27,7 @@
 
     List<VersionQueryMetric> getAll();
 
-    void record(String groupId, String artifactId, String versionId);
+    void insert(VersionQueryMetric versionQueryMetric);
```
Comment (elopezcastro) [inline legend-depot-store-api/src/main/java/org/finos/legend/depot/store/admin/api/metrics/QueryMetricsStore.java:30]: and as per previous comment lets move any metrics related to metrics mocule, including this

**Review by elopezcastro — CHANGES_REQUESTED** [review]

File: legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/services/QueryMetricsHandler.java (hunk 5 above)
Comment (gs-gunjan) [reply legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/services/QueryMetricsHandler.java:78]: the underlying implementation is a concurrent queue, and it uses the FIFO rather than an iterator which needs a locking mech, so even if the queue is being modified, it depends which operation completed first the polling or the recording which would decide whether this comes out to be true or not

File: legend-depot-store-api/src/main/java/org/finos/legend/depot/store/admin/api/metrics/QueryMetricsStore.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (gs-gunjan) [reply legend-depot-store-api/src/main/java/org/finos/legend/depot/store/admin/api/metrics/QueryMetricsStore.java:30]: this is the store api, this actually inserts the metric in the db
the container should have the record naming

File: legend-depot-core-services/pom.xml [hunk 8]
```diff
@@ -78,6 +82,10 @@
             <artifactId>mongo-java-server</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
Comment (elopezcastro) [inline legend-depot-core-services/pom.xml:85]: the dependency is duplicated see above

File: legend-depot-core-services/src/main/java/org/finos/legend/depot/services/ReadDataServicesModule.java [hunk 9]
```diff
@@ -41,4 +46,12 @@ protected void configure()
         expose(EntityClassifierService.class);
         expose(FileGenerationsService.class);
     }
+
+    @Provides
+    @Named("queryMetricsRegistry")
```
Comment (elopezcastro) [inline legend-depot-core-services/src/main/java/org/finos/legend/depot/services/ReadDataServicesModule.java:51]: this should be in the metrics private module

File: legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/MetricsModule.java [hunk 10]
```diff
@@ -20,21 +20,26 @@
...
 import org.finos.legend.depot.store.metrics.services.QueryMetricsHandler;
+import org.finos.legend.depot.store.metrics.services.InMemoryQueryMetricsRegistry;
 
 public class MetricsModule extends PrivateModule
 {
     @Override
     protected void configure()
     {
         bind(QueryMetricsHandler.class);
+        bind(QueryMetricsRegistry.class).to(InMemoryQueryMetricsRegistry.class);
```
Comment (elopezcastro) [inline legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/MetricsModule.java:33]: you dont need this... move the intializtion from services to this module ...u dont need this bind

File: legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/MetricsModule.java [hunk 11]
```diff
@@ -20,21 +20,26 @@
...
+
         expose(QueryMetricsHandler.class);
+        expose(QueryMetricsRegistry.class);
```
Comment (elopezcastro) [inline legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/MetricsModule.java:36]: same as oer above...move the innit you have in services module

File: legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/store/api/QueryMetricsStore.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
+//  Copyright 2021 Goldman Sachs
+//
```
Comment (elopezcastro) [inline legend-depot-store-metrics/src/main/java/org/finos/legend/depot/store/metrics/store/api/QueryMetricsStore.java:2]: do we need two sepaate interfaces... merge unto one and call ins QueryMetricsStore

**Review by elopezcastro — CHANGES_REQUESTED** [review]

**Review by elopezcastro — APPROVED** [review]
